cli/activate.go

```diff
@@ -1,138 +1,188 @@
 	if err != nil {
-		log.Fatalf("Failed to read steps spec (spec.json), error: %s", err)
+		return fmt.Errorf("activate: failed to read %s steplib: %s", libraryURL, err)
 	}
 
-	_, stepFound, versionFound := collection.GetStep(id, version)
-	if !stepFound || !versionFound {
-		if !update {
-			if !stepFound {
-				log.Fatalf("Collection doesn't contain step with id: %s", id)
-			} else if !versionFound {
-				log.Fatalf("Collection doesn't contain step (%s) with version: %s", id, version)
-			}
-		}
+	step, version, err := queryStep(library, id, version, updateLibrary)
+	if err != nil {
+		return fmt.Errorf("activate: failed to find step: %s", err)
+	}
+
+	srcFolder, err := downloadStep(library, id, version, step)
+	if err != nil {
+		return fmt.Errorf("activate: failed to download step: %s", err)
+	}
 
-		if !stepFound {
-			log.Infof("Collection doesn't contain step with id: %s -- Updating StepLib", id)
-		} else if !versionFound {
-			log.Infof("Collection doesn't contain step (%s) with version: %s -- Updating StepLib", id, version)
+	if err := copyStep(srcFolder, destination); err != nil {
+		return fmt.Errorf("activate: copy step failed: %s", err)
+	}
+
+	if destinationStepYML != "" {
+		if err := copyStepYML(libraryURL, id, version, destinationStepYML); err != nil {
+			return fmt.Errorf("activate: copy step.yml failed: %s", err)
 		}
+	}
+
+	return nil
+}
+
+func queryStep(library models.StepCollectionModel, id, version string, updateLibrary bool) (models.StepModel, string, error) {
+	step, stepFound, versionFound := library.GetStep(id, version)
+	if (!stepFound || !versionFound) && updateLibrary {
+		log.Infof("StepLib doesn't contain step (%s) with version: %s -- Updating StepLib", id, version)
 
-		collection, err = stepman.UpdateLibrary(collectionURI)
+		var err error
+		library, err = stepman.UpdateLibrary(library.SteplibSource)
 		if err != nil {
-			log.Fatalf("Failed to update collection (%s), err: %s", collectionURI, err)
+			return models.StepModel{}, "", fmt.Errorf("query step: failed to update %s steplib: %s", library.SteplibSource, err)
 		}
 
-		_, stepFound, versionFound := collection.GetStep(id, version)
-		if !stepFound {
-			if !stepFound {
-				log.Fatalf("Collection doesn't contain step with id: %s", id)
-			} else if !versionFound {
-				log.Fatalf("Collection doesn't contain step (%s) with version: %s", id, version)
-			}
-		}
+		step, stepFound, versionFound = library.GetStep(id, version)
+	}
+	if !stepFound {
+		return models.StepModel{}, "", fmt.Errorf("query step: %s steplib does not contain %s step", library.SteplibSource, id)
+	}
+	if !versionFound {
+		return models.StepModel{}, "", fmt.Errorf("query step: %s steplib does not contain %s step %s version", library.SteplibSource, id, version)
 	}
 
-	// If version doesn't provided use latest
 	if version == "" {
-		latest, err := collection.GetLatestStepVersion(id)
+		latest, err := library.GetLatestStepVersion(id)
```

Comment:
This seems to related to the removed ` Check step exist in local cache` part. It may be removed as the GetStep function also handles version is empty string case.

Reply:
later we need the exact version: 
- https://github.com/bitrise-io/stepman/pull/283/files#diff-6e1b0a2810695a1f8a5122e23f318696R88
- https://github.com/bitrise-io/stepman/pull/283/files#diff-6e1b0a2810695a1f8a5122e23f318696R98